flink-table/flink-table-common/src/main/java/org/apache/flink/table/data/DecimalData.java

```diff
@@ -0,0 +1,260 @@
+	// in order to be accessible for DecimalDataUtils
+
+	static final int MAX_COMPACT_PRECISION = 18;
+
+	/**
+	 * Maximum number of decimal digits an Int can represent. (1e9 < Int.MaxValue < 1e10)
+	 */
+	static final int MAX_INT_DIGITS = 9;
+
+	/**
+	 * Maximum number of decimal digits a Long can represent. (1e18 < Long.MaxValue < 1e19)
+	 */
+	static final int MAX_LONG_DIGITS = 18;
+
+	static final long[] POW10 = new long[MAX_COMPACT_PRECISION + 1];
+
+	static {
+		POW10[0] = 1;
+		for (int i = 1; i < POW10.length; i++) {
+			POW10[i] = 10 * POW10[i - 1];
+		}
+	}
+
+	// The semantics of the fields are as follows:
+	//  - `precision` and `scale` represent the precision and scale of SQL decimal type
+	//  - If `decimalVal` is set, it represents the whole decimal value
+	//  - Otherwise, the decimal value is longVal/(10^scale).
+	//
+	// Note that the (precision, scale) must be correct.
+	// if precision > MAX_COMPACT_PRECISION,
+	//   `decimalVal` represents the value. `longVal` is undefined
+	// otherwise, (longVal, scale) represents the value
+	//   `decimalVal` may be set and cached
+
+	final int precision;
+	final int scale;
+
+	final long longVal;
+	BigDecimal decimalVal;
+
+	// this constructor does not perform any sanity check.
+	DecimalData(int precision, int scale, long longVal, BigDecimal decimalVal) {
+		this.precision = precision;
+		this.scale = scale;
+		this.longVal = longVal;
+		this.decimalVal = decimalVal;
+	}
+
+	// ------------------------------------------------------------------------------------------
+	// Public Interfaces
+	// ------------------------------------------------------------------------------------------
+
+	/**
+	 * Returns the <i>precision</i> of this {@link DecimalData}.
+	 *
+	 * <p>The precision is the number of digits in the unscaled value.
+	 */
+	public int precision() {
+		return precision;
+	}
+
+	/**
+	 * Returns the <i>scale</i> of this {@link DecimalData}.
+	 */
+	public int scale() {
+		return scale;
+	}
+
+	/**
+	 * Converts this {@link DecimalData} into an instance of {@link BigDecimal}.
+	 */
+	public BigDecimal toBigDecimal() {
+		BigDecimal bd = decimalVal;
+		if (bd == null) {
+			decimalVal = bd = BigDecimal.valueOf(longVal, scale);
+		}
+		return bd;
+	}
+
+	/**
+	 * Returns a long describing the <i>unscaled value</i> of this {@link DecimalData}.
+	 *
+	 * @throws ArithmeticException if this {@link DecimalData} does not exactly fit in a long.
+	 */
+	public long toUnscaledLong() {
+		if (isCompact()) {
+			return longVal;
+		} else {
+			return toBigDecimal().unscaledValue().longValueExact();
+		}
+	}
+
+	/**
+	 * Returns a byte array describing the <i>unscaled value</i> of this {@link DecimalData}.
+	 *
+	 * @return the unscaled byte array of this {@link DecimalData}.
+	 */
+	public byte[] toUnscaledBytes() {
+		if (!isCompact()) {
+			return toBigDecimal().unscaledValue().toByteArray();
+		}
+
+		// big endian; consistent with BigInteger.toByteArray()
+		byte[] bytes = new byte[8];
```

Comment:
Hi @wuchong , this is a wrong implementation here. Can you take https://issues.apache.org/jira/browse/FLINK-16922 and fix that?

Reply:
OK. Will do that in FLINK-16996.